src/components/Modal/index.tsx

```diff
@@ -2,39 +2,62 @@ import React from 'react';
 import { _cs } from '@togglecorp/fujs';
 import { IoClose } from 'react-icons/io5';
 
+import ElementFragments from '../ElementFragments';
 import Button from '../Button';
 import BodyBackdrop from '../BodyBackdrop';
 
 import styles from './styles.css';
 
-export interface Props{
+interface BaseProps {
     children?: React.ReactNode;
     heading?: React.ReactNode;
     footer?: React.ReactNode;
     className?: string;
     bodyClassName?: string;
-    headingClassName?: string;
+    headerClassName?: string;
     footerClassName?: string;
-    onClose: () => void;
-    closeButtonHidden?: boolean;
+    headerIcons?: React.ReactNode;
+    headerActions?: React.ReactNode;
+    footerIcons?: React.ReactNode;
+    footerActions?: React.ReactNode;
 }
 
+export type Props = BaseProps & ({
+    hideCloseButton?: false;
+    onCloseButtonClick: () => void;
+} | {
+    hideCloseButton: true;
+    onCloseButtonClick?: never;
+})
+
+
+// # New breaking changes
+// - onClose ==> onCloseButtonClick
+// - headingClassName ==> headerClassName
+// - sending heading = null no longer hides header (instead it's hidden if none of header elements are provided and hideCloseButton is set)
+
 function Modal(props: Props) {
     const {
         heading,
         children,
         footer,
 
         className,
-        headingClassName,
+        headerClassName,
         bodyClassName,
         footerClassName,
 
-        onClose,
+        footerActions,
+        footerIcons,
+        headerActions,
+        headerIcons,
 
-        closeButtonHidden,
+        onCloseButtonClick,
+        hideCloseButton,
     } = props;
 
+    const shouldHideHeader = hideCloseButton && !heading && !headerActions && !headerIcons;
```

Comment:
This should be independent meaning shouldHideHeader should always hide header irrespective of whether or not heading items are provided as this way its use can be extended and also will be consistent with other components.